### ISSUE TITLE
iRODS 4.2.10 replication of a file with a space fails in a compound resource using msisync_to_archive

### ISSUE THREAD

**Issue by cookie33** [irods/irods#6029]
- [x] main
- [x] 4-2-stable
-------

### iRODS Version, OS and Version

iRODS 4.2.10 Centos 7.9

### What did you try to do?

put a file with a space in the filename in a compound resource.
Use a delayed rule to replicate to the archive resource,
The replication fails because the file is not replicated to the archive resource.

### Expected behavior

The file is replicated to the archive resource.

### Observed behavior (including steps to reproduce, if applicable)

# put file in compound resource
```
iput -R eudat univMSSInterface.sh "test 1.txt"
```
# show file in compound resource
```
ils -l
/igor/home/rods:
  rods              0 eudat;eudatCache         5204 2021-12-02.09:58 & test 1.txt
...
```
# there is a delayed rule to replicate the file:
```
iqstat -al
Pending rule-executions
id: 25410
name:
                        *CompoundRescName="eudat"
                        *CacheRescName   ="*CompoundRescName;eudatCache";
                        *ArchiveRescName ="*CompoundRescName;eudatPnfs";
                        writeLine("serverLog","Execute command to replicate (in resource *CompoundRescName) *obj_path (*file_path) to *ArchiveRescName, because of put");
                        msisync_to_archive("*CacheRescName", "*file_path", "*obj_path" );

rei_file_path: EMPTY_REI_PATH
user_name: rods
address:
time: 1638435553 : 2021-12-02.09:59:13
frequency: 1h DOUBLE UNTIL SUCCESS OR 6 TIMES
priority: 5
estimated_exe_time:
notification_addr:
last_exe_time:
exec_status:
```
# the file is not properly replicated. Notice the X
```
ils -l
/igor/home/rods:
  rods              0 eudat;eudatCache         5204 2021-12-02.09:58 & test 1.txt
  rods              1 eudat;eudatPnfs         5204 2021-12-02.09:58 X test 1.txt
...
```
# there are errors in the logfile of iRODS
```
Dec  2 09:59:20 pid:9952 NOTICE: writeLine: inString = Execute command to replicate (in resource eudat) /igor/home/rods/test 1.txt (/data/eudatCache/Vault/home/rods/test 1.txt) to eudat;eudatPnfs, because of put
Dec  2 09:59:20 pid:9955 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:mkdir '/data/eudatPnfs/home/rods'
Dec  2 09:59:20 pid:9958 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:chmod '/data/eudatPnfs/home/rods' 750
Dec  2 09:59:20 pid:9961 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:syncToArch /data/eudatCache/Vault/home/rods/test 1.txt /data/eudatPnfs/home/rods/test 1.txt
Dec  2 09:59:20 pid:9964 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:chmod '/data/eudatPnfs/home/rods/test 1.txt' 600
Dec  2 09:59:20 pid:9952 remote addresses: 127.0.0.1, x.x.x.x ERROR: _rsExecCmd: waitpid status = 9964, myExecCmdOut->status = 0, childStatus = 256
Dec  2 09:59:20 pid:9952 WARNING: No replica access token in L1 descriptor. Ignoring replica access table. [path=/igor/home/rods/test 1.txt, resource_hierarchy=eudat;eudatPnfs]
Dec  2 09:59:20 pid:9967 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:stat '/data/eudatPnfs/home/rods/test 1.txt'
Dec  2 09:59:20 pid:9952 remote addresses: 127.0.0.1, x.x.x.x ERROR: _rsExecCmd: waitpid status = 9967, myExecCmdOut->status = 0, childStatus = 256
Dec  2 09:59:20 pid:9952 remote addresses: 127.0.0.1, x.x.x.x ERROR: [get_size_in_vault:159] - getSizeInVault error [error_code=[-555107], path=[/igor/home/rods/test 1.txt], hierarchy=[eudat;eudatPnfs]]
Dec  2 09:59:20 pid:9952 remote addresses: 127.0.0.1, x.x.x.x ERROR: [rsDataObjClose:794] - [UNIV_MSS_STAT_ERR: [update_replica_size_and_throw_on_failure:430] - failed to get size in vault [error_code=[-555107], path=[/igor/home/rods/test 1.txt], hierarchy=[eudat;eudatPnfs]]

]
Dec  2 09:59:20 pid:9952 remote addresses: 127.0.0.1, 145.100.3.239 ERROR: [close_replica] - rsDataObjClose failed with [-555107]
```

# The delayed rule is as follows:
```
                *file_path=$filePath;
                *obj_path=$objPath;
                delay("<PLUSET>1m</PLUSET><INST_NAME>irods_rule_engine_plugin-irods_rule_language-instance</INST_NAME><EF>1h DOUBLE UNTIL SUCCESS OR 6 TIMES</EF>") {
                        *CompoundRescName="eudat"
                        *CacheRescName   ="*CompoundRescName;eudatCache";
                        *ArchiveRescName ="*CompoundRescName;eudatPnfs";
                        writeLine("serverLog","Execute command to replicate (in resource *CompoundRescName) *obj_path (*file_path) to *ArchiveRescName, because of put");
                        msisync_to_archive("*CacheRescName", "*file_path", "*obj_path" );
```

**Comment by cookie33** [comment]
I think this comes from the fact that the `filepath/filename` parameters for the `syncToArch` function are NOT between quotes. So it sees multiple items when trying to do:
```
syncToArch /data/eudatCache/Vault/home/rods/test 1.txt /data/eudatPnfs/home/rods/test 1.txt
```
Instead of three parameters It are five parameters.

**Comment by trel** [comment]
I believe this is the line that needs more single quotes...

https://github.com/irods/irods/blob/36030a029b4a6d271095383237b0676f2a83d0d5/plugins/resources/univmss/libunivmss.cpp#L624

The single quotes were there in March 2013... 
https://github.com/irods/irods/blame/4746e05fe067da8366803dbfcff906356bf31a9d/plugins/resources/univmss/libunivmss.cpp#L667

but removed in May 2013 in the move to the plugin architecture...
https://github.com/irods/irods/blame/3a0a1275dd3e2781dfe3523da345726c169e613b/plugins/resources/univmss/libunivmss.cpp#L662

in this commit
https://github.com/irods/irods/commit/856cd7030d999b50f09502fa4c9b87deb2cfb868

--

We'll get the single quotes back in there.

**Comment by trel** [comment]
26 months ago... https://github.com/irods/irods/pull/4554

**Comment by cookie33** [comment]
I am also busy testing this with the `univMSSInterface.sh` to see if changes are needed in here to get it to work. This also seems not to work. Maybe that is the problem.

**Comment by cookie33** [comment]
Checked the version active and the version from a template:
```
diff -b univMSSInterface.sh univMSSInterface.sh.template
8c8
< # Your working version should be in this directory server/bin/cmd/univMSSInterface.sh.
---
> # Your working version should be in this directory msiExecCmd_bin/univMSSInterface.sh.
18c18
<         op=`which cp`
---
>         op=`which template-cp`
28c28
<         op=`which cp`
---
>         op=`which template-cp`
38c38
<         op=`which mkdir`
---
>         op=`which template-mkdir`
53c53
<         op=`which chmod`
---
>         op=`which template-chmod`
62,63c62,63
<     op=`which rm`
<       `$op -f $1`
---
>       op=`which template-rm`
>       `$op $1`
71c71
<     op=`which mv`
---
>     op=`which template-mv`
78c78
<         op=`which stat`
---
>         op=`which template-stat`
```

The only difference seems to be the template for each command. And I added a `-f` flag in the remove of a file

testing a file with a space in it:
```
./univMSSInterface.sh stat "/data/eudatPnfs/home/sarar#frank/Data Archive v1.1.png"
/bin/stat: cannot stat ‘/data/eudatPnfs/home/sarar#frank/Data’: No such file or directory
```

Test a file without a space in it:
```
./univMSSInterface.sh stat /data/eudatPnfs/home/robertv/test_data.txt
fd00h/64768d:5003681:0600:1:38512:35390:0:12:4096:8:2019-02-11-14.28.40:2020-05-14-11.26.52:2020-05-14-11.26.52
```

So the supplied template of `univMSSInterface.sh` also does not seem to support files with spaces in it.

So it might be two problems:
 * univMSSInterface.sh.template does not support files with spaces
 * the  msisync_to_archive for syncToArch does not have quotes around the parameters

**Comment by cookie33** [comment]
Modified local `univMSSInterface.sh `:
```
diff -b univMSSInterface.sh univMSSInterface.sh.template
8c8
< # Your working version should be in this directory server/bin/cmd/univMSSInterface.sh.
---
> # Your working version should be in this directory msiExecCmd_bin/univMSSInterface.sh.
18,19c18,19
<         op=`which cp`
<         `$op "$1" "$2"`
---
>         op=`which template-cp`
>         `$op $1 $2`
28,29c28,29
<         op=`which cp`
<         `$op "$1" "$2"`
---
>         op=`which template-cp`
>         `$op $1 $2`
38,39c38,39
<         op=`which mkdir`
<         `$op -p "$1"`
---
>         op=`which template-mkdir`
>         `$op -p $1`
53,54c53,54
<         op=`which chmod`
<         `$op $2 "$1"`
---
>         op=`which template-chmod`
>         `$op $2 $1`
62,63c62,63
<     op=`which rm`
<       `$op -f "$1"`
---
>       op=`which template-rm`
>       `$op $1`
71,72c71,72
<     op=`which mv`
<     `$op "$1" "$2"`
---
>     op=`which template-mv`
>     `$op $1 $2`
78,79c78,79
<         op=`which stat`
<       output=`$op "$1"`
---
>         op=`which template-stat`
>       output=`$op $1`
125,131c125,131
<       syncToArch ) $1 "$2" "$3" ;;
<       stageToCache ) $1 "$2" "$3" ;;
<       mkdir ) $1 "$2" ;;
<       chmod ) $1 "$2" "$3" ;;
<       rm ) $1 "$2" ;;
<       mv ) $1 "$2" "$3" ;;
<       stat ) $1 "$2" ;;
---
>       syncToArch ) $1 $2 $3 ;;
>       stageToCache ) $1 $2 $3 ;;
>       mkdir ) $1 $2 ;;
>       chmod ) $1 $2 $3 ;;
>       rm ) $1 $2 ;;
>       mv ) $1 $2 $3 ;;
>       stat ) $1 $2 ;;
```

test again a file with a space:
```
./univMSSInterface.sh stat "/data/eudatPnfs/home/sarar#frank/Data Archive v1.1.png"
fd00h/64768d:8435544:0600:1:38512:35390:0:256433:4096:504:2021-12-01-15.59.46:2021-12-01-15.59.46:2021-12-01-16.01.47
```
test a file without a space:
```
./univMSSInterface.sh stat /data/eudatPnfs/home/robertv/test_data.txt
fd00h/64768d:5003681:0600:1:38512:35390:0:12:4096:8:2019-02-11-14.28.40:2020-05-14-11.26.52:2020-05-14-11.26.52
```
Test copy of a file with a space:
```
./univMSSInterface.sh syncToArch '/data/eudatCache/Vault/home/rods/test 4.txt' '/data/eudatPnfs/home/rods/test 4.txt'
UNIVMSS /bin/cp /data/eudatCache/Vault/home/rods/test 4.txt /data/eudatPnfs/home/rods/test 4.txt
```
This now works..

Try again In iRODS:
```
iput -R eudat univMSSInterface.sh "test 5.txt"
```
It is there:
```
ils -l
/igor/home/rods:
...
  rods              0 eudat;eudatCache         5246 2021-12-03.14:46 & test 5.txt
...
```
The copy using the msi still fails:
```
ils -l
/igor/home/rods:
...
  rods              0 eudat;eudatCache         5246 2021-12-03.14:46 & test 5.txt
  rods              1 eudat;eudatPnfs         5246 2021-12-03.14:46 X test 5.txt
...
```

And the error still is:
```
Dec  3 14:47:58 pid:29002 NOTICE: writeLine: inString = Execute command to replicate (in resource eudat) /igor/home/rods/test 5.txt (/data/eudatCache/Vault/home/rods/test 5.txt) to eudat;eudatPnfs, because of put
Dec  3 14:47:59 pid:29005 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:mkdir '/data/eudatPnfs/home/rods'
Dec  3 14:47:59 pid:29008 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:chmod '/data/eudatPnfs/home/rods' 750
Dec  3 14:47:59 pid:29011 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:syncToArch /data/eudatCache/Vault/home/rods/test 5.txt /data/eudatPnfs/home/rods/test 5.txt
Dec  3 14:47:59 pid:29014 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:chmod '/data/eudatPnfs/home/rods/test 5.txt' 600
Dec  3 14:47:59 pid:29002 remote addresses: 127.0.0.1, x.x.x.x ERROR: _rsExecCmd: waitpid status = 29014, myExecCmdOut->status = 0, childStatus = 256
Dec  3 14:47:59 pid:29002 WARNING: No replica access token in L1 descriptor. Ignoring replica access table. [path=/igor/home/rods/test 5.txt, resource_hierarchy=eudat;eudatPnfs]
Dec  3 14:47:59 pid:29017 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:stat '/data/eudatPnfs/home/rods/test 5.txt'
Dec  3 14:47:59 pid:29002 remote addresses: 127.0.0.1, x.x.x.x ERROR: _rsExecCmd: waitpid status = 29017, myExecCmdOut->status = 0, childStatus = 256
Dec  3 14:47:59 pid:29002 remote addresses: 127.0.0.1, x.x.x.x ERROR: [get_size_in_vault:159] - getSizeInVault error [error_code=[-555107], path=[/igor/home/rods/test 5.txt], hierarchy=[eudat;eudatPnfs]]
Dec  3 14:47:59 pid:29002 remote addresses: 127.0.0.1, x.x.x.x ERROR: [rsDataObjClose:794] - [UNIV_MSS_STAT_ERR: [update_replica_size_and_throw_on_failure:430] - failed to get size in vault [error_code=[-555107], path=[/igor/home/rods/test 5.txt], hierarchy=[eudat;eudatPnfs]]

]
Dec  3 14:47:59 pid:29002 remote addresses: 127.0.0.1, x.x.x.x ERROR: [close_replica] - rsDataObjClose failed with [-555107]
```

So there are two problems:
* default univMSSInterface.sh.template does not work with files with spaces in the filename
* the `msisync_to_archive` fails to copy because of missing quotes

**Comment by trel** [comment]
confirmed/reproduced.

with this diff...

```diff
~/msiExecCmd_bin$ diff univMSSInterface.sh.template univMSSInterface.sh
79c79
<       output=`$op $1`
---
>       output=`$op "$1"`
131c131
<       stat ) $1 $2 ;;
---
>       stat ) "$1" "$2" ;;
```

i see this success...

```
~/msiExecCmd_bin$ ./univMSSInterface.sh stat '/tmp/univmsstest space.txt'
801h/2049d:27834285:0664:1:999:999:0:5:4096:8:2021-12-03-08.43.43:2021-12-03-08.43.43:2021-12-03-08.43.43
```

so yes, there probably needs to be more protection for the other commands in there as well.

**Comment by cookie33** [comment]
One other observation. The univMSSInterface.sh.template has the following two functions with echo statements:
```
syncToArch () {
        # <your command or script to copy from cache to MSS> $1 $2
        # e.g: /usr/local/bin/rfcp $1 rfioServerFoo:$2
        op=`which template-cp`
        `$op "$1" "$2"`
        echo "UNIVMSS $op \"$1\" \"$2\""
        return
}

# function for staging a file $1 from the MSS to file $2 on disk
stageToCache () {
        # <your command to stage from MSS to cache> $1 $2
        # e.g: /usr/local/bin/rfcp rfioServerFoo:$1 $2
        op=`which template-cp`
        `$op "$1" "$2"`
        echo "UNIVMSS $op \"$1\" \"$2\""
        return
}
```

If the echo command is put in an other place the function will give back errors if there is an error:
```
syncToArch () {
        # <your command or script to copy from cache to MSS> $1 $2
        # e.g: /usr/local/bin/rfcp $1 rfioServerFoo:$2
        echo "UNIVMSS $op \"$1\" \"$2\""
        op=`which cp`
        `$op "$1" "$2"`
        return
}

# function for staging a file $1 from the MSS to file $2 on disk
stageToCache () {
        # <your command to stage from MSS to cache> $1 $2
        # e.g: /usr/local/bin/rfcp rfioServerFoo:$1 $2
        echo "UNIVMSS $op \"$1\" \"$2\""
        op=`which cp`
        `$op "$1" "$2"`
        return
}
```

with the template (echo command on standard place):
```
./univMSSInterface.sh.template syncToArch '/data/eudatCache/Vault/home/rods/test 55.txt' '/data/eudatPnfs/home/rods/test 5.txt'
which: no template-cp in (/usr/sara/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/rods/.local/bin:/home/rods/bin)
./univMSSInterface.sh.template: line 19: /data/eudatCache/Vault/home/rods/test 55.txt: No such file or directory
UNIVMSS  "/data/eudatCache/Vault/home/rods/test 55.txt" "/data/eudatPnfs/home/rods/test 5.txt"

echo $?
0
```
With a modified file (echo command before actions) :
```
./univMSSInterface.sh syncToArch '/data/eudatCache/Vault/home/rods/test 55.txt' '/data/eudatPnfs/home/rods/test 5.txt'
UNIVMSS  "/data/eudatCache/Vault/home/rods/test 55.txt" "/data/eudatPnfs/home/rods/test 5.txt"
/bin/cp: cannot stat ‘/data/eudatCache/Vault/home/rods/test 55.txt’: No such file or directory

echo $?
1
```

**Comment by trel** [comment]
Ah, that does seem important as the other machinery depends on the error codes.   Very good, we can update the template with that reordering as well.

**Comment by cookie33** [comment]
This also explains why there never was an error during the copy from and to the archive. The echo command went fine. So the error was lost. with the echo on an other place the error is shown:
```
Dec  3 15:28:35 pid:750 NOTICE: writeLine: inString = Execute command to replicate (in resource eudat) /igor/home/rods/test 6.txt (/data/eudatCache/Vault/home/rods/test 6.txt) to eudat;eudatPnfs, because of put
Dec  3 15:28:35 pid:753 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:mkdir '/data/eudatPnfs/home/rods'
Dec  3 15:28:35 pid:756 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:chmod '/data/eudatPnfs/home/rods' 750
Dec  3 15:28:35 pid:759 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:syncToArch /data/eudatCache/Vault/home/rods/test 6.txt /data/eudatPnfs/home/rods/test 6.txt
Dec  3 15:28:35 pid:750 remote addresses: 127.0.0.1, x.x.x.x ERROR: _rsExecCmd: waitpid status = 759, myExecCmdOut->status = 0, childStatus = 256
Dec  3 15:28:35 pid:750 remote addresses: 127.0.0.1, x.x.x.x ERROR: [-]   /irods/server/api/src/rsFileSyncToArch.cpp:191:int _rsFileSyncToArch(rsComm_t *, fileStageSyncInp_t *, fileSyncOut_t **) :  status [UNIV_MSS_SYNCTOARCH_ERR]  errno [Transport endpoint is not connected] -- message [fileSyncToArch failed for [/data/eudatPnfs/home/rods/test 6.txt]]
        [-]     /irods/server/drivers/src/fileDriver.cpp:612:irods::error fileSyncToArch(rsComm_t *, irods::first_class_object_ptr, const std::string &) :  status [UNIV_MSS_SYNCTOARCH_ERR]  errno [Transport endpoint is not connected] -- message [failed to call 'synctoarch']
                [-]     /irods/plugins/resources/univmss/libunivmss.cpp:655:irods::error univ_mss_file_sync_to_arch(irods::plugin_context &, const char *) :  status [UNIV_MSS_SYNCTOARCH_ERR]  errno [Transport endpoint is not connected] -- message [univ_mss_file_sync_to_arch: copy of [/data/eudatCache/Vault/home/rods/test 6.txt] to [/data/eudatPnfs/home/rods/test 6.txt] failed.   stdout buff [0x164e390]   stderr buff [0x1682040]  status [-344000]]

Dec  3 15:28:35 pid:750 WARNING: No replica access token in L1 descriptor. Ignoring replica access table. [path=/igor/home/rods/test 6.txt, resource_hierarchy=eudat;eudatPnfs]
Dec  3 15:28:35 pid:762 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/univMSSInterface.sh argv:stat '/data/eudatPnfs/home/rods/test 6.txt'
Dec  3 15:28:35 pid:750 remote addresses: 127.0.0.1, x.x.x.x ERROR: _rsExecCmd: waitpid status = 762, myExecCmdOut->status = 0, childStatus = 256
Dec  3 15:28:35 pid:750 remote addresses: 127.0.0.1, x.x.x.x ERROR: [get_size_in_vault:159] - getSizeInVault error [error_code=[-555107], path=[/igor/home/rods/test 6.txt], hierarchy=[eudat;eudatPnfs]]
Dec  3 15:28:35 pid:750 remote addresses: 127.0.0.1, x.x.x.x ERROR: [rsDataObjClose:794] - [UNIV_MSS_STAT_ERR: [update_replica_size_and_throw_on_failure:430] - failed to get size in vault [error_code=[-555107], path=[/igor/home/rods/test 6.txt], hierarchy=[eudat;eudatPnfs]]

]
Dec  3 15:28:35 pid:750 remote addresses: 127.0.0.1, x.x.x.x ERROR: [close_replica] - rsDataObjClose failed with [-555107]
```

**Comment by trel** [comment]
I am hoping that fixing the 'error code getting stomped' also fixes #6030... will need a test to confirm this.

**Comment by trel** [comment]
confirmed - correctly sending a non-zero return code keeps the failed rule in the queue to be tried again.
